Added operational model for atoi and atol

lucasccordeiro opened the pull request:
This PR adds an operational model for the `atoi` function. It should fix the issue https://github.com/esbmc/esbmc/issues/1009.

lucasccordeiro committed:
Added operational model for atoi

Added operational model for the C function atoi.

ChenfengWei0 commented on regression/esbmc/github_1009/main.c:
```diff
@@ -0,0 +1,28 @@
+  assert(result == 2147483647);
+  
+  return 0;
+}
```
I think we can also have an assertion for the empty string:

```c++
result = atoi("");
assert(result == 0);

char* s = NULL;
result = atoi(s);   // report undefined behavior? Might cause segfault
```

lucasccordeiro replied:
Implemented as suggested,

ChenfengWei0 reviewed (approved):
LGTM

lucasccordeiro committed:
Added regression tests for atoi

Signed-off-by: Lucas Cordeiro <lucasccordeiro@gmail.com>

lucasccordeiro commented:
Thanks for your comments, @ChenfengWei0.

kunjsong01 commented:
Looks good. 
Just a minor remark for a possible simplification (probably in another PR): 
Could it be further simplified to a casting of `strtol`? 
```c
(int)strtol(nptr, (char **)NULL, 10)
```
so that we have less code to maintain in the future. 

C11 draft N1570 §7.22.1.2:
```
Except for the behavior on error, they are equivalent to
atoi: (int)strtol(nptr, (char **)NULL, 10)
atol: strtol(nptr, (char **)NULL, 10)
atoll: strtoll(nptr, (char **)NULL, 10)
```

lucasccordeiro committed:
Added regression test for atoi

Signed-off-by: Lucas Cordeiro <lucasccordeiro@gmail.com>

lucasccordeiro committed:
Simplify atoi as a casting of strtol

Simplified the atoi function as a casting of strtol:

(int)strtol(nptr, (char **)NULL, 10)

C11 draft N1570 §7.22.1.2:
Except for the behavior on error, they are equivalent to
atoi: (int)strtol(nptr, (char **)NULL, 10)
atol: strtol(nptr, (char **)NULL, 10)
atoll: strtoll(nptr, (char **)NULL, 10)

Signed-off-by: Lucas Cordeiro <lucasccordeiro@gmail.com>

lucasccordeiro commented:
@kunjsong01: I have addressed your above comment in this PR. 
CBMC is already following exactly what you suggested above. 
So, I have updated this PR to follow the same approach.

lucasccordeiro committed:
Added atol operational model

Signed-off-by: Lucas Cordeiro <lucasccordeiro@gmail.com>

lucasccordeiro committed:
Added test cases for atol

lucasccordeiro commented:
@kunjsong01: 

I have created this issue https://github.com/esbmc/esbmc/issues/1016 to handle the operational models for `atoll` and `strtoll`.

I'm now running this PR over the SV-COMP benchmarks. I'll post the results here once this run is ready.

Many thanks for your comments.

lucasccordeiro commented:
This atoi model hugely affects the ESBMC performance over the Juliet benchmarks.

This PR:

````
2023-05-14T11:22:21.7611290Z Statistics:          26459 Files
2023-05-14T11:22:21.7613172Z   correct:           13998
2023-05-14T11:22:21.7614019Z     correct true:     7788
2023-05-14T11:22:21.7614597Z     correct false:    6210
2023-05-14T11:22:21.7615124Z   incorrect:            88
2023-05-14T11:22:21.7615615Z     incorrect true:     39
2023-05-14T11:22:21.7628548Z     incorrect false:    49
2023-05-14T11:22:21.7630732Z   unknown:           12373
2023-05-14T11:22:21.7649566Z   Score:             19754 (max: 43043)
````

Master:
````
Statistics:          26459 Files
   correct:            14488
     correct true:     8269
     correct false:    6219
   incorrect:            88
     incorrect true:     39
    incorrect false:     49
   unknown:            11883
   Score:              20725 (max: 43043)
````

lucasccordeiro committed:
try alternative operational model for strtol

Signed-off-by: Lucas Cordeiro <lucasccordeiro@gmail.com>

lucasccordeiro commented:
The results of the alternative operational model:

````
2023-05-15T07:42:07.5195484Z Statistics:          26459 Files
2023-05-15T07:42:07.5196306Z   correct:           14037
2023-05-15T07:42:07.5196863Z     correct true:     7829
2023-05-15T07:42:07.5239618Z     correct false:    6208
2023-05-15T07:42:07.5241658Z   incorrect:            88
2023-05-15T07:42:07.5243241Z     incorrect true:     39
2023-05-15T07:42:07.5243920Z     incorrect false:    49
2023-05-15T07:42:07.5244594Z   unknown:           12334
2023-05-15T07:42:07.5245120Z   Score:             19834 (max: 43043)
````

lucasccordeiro committed:
try again an alternative operational model

Signed-off-by: Lucas Cordeiro <lucasccordeiro@gmail.com>

lucasccordeiro committed:
clean up the OM

lucasccordeiro commented:
Results of the second attempt for the OM:

````
2023-05-15T21:10:52.3806963Z Statistics:          26459 Files
2023-05-15T21:10:52.3808086Z   correct:           14025
2023-05-15T21:10:52.3809120Z     correct true:     7815
2023-05-15T21:10:52.3842851Z     correct false:    6210
2023-05-15T21:10:52.3843890Z   incorrect:            88
2023-05-15T21:10:52.3844474Z     incorrect true:     39
2023-05-15T21:10:52.3845025Z     incorrect false:    49
2023-05-15T21:10:52.3845703Z   unknown:           12346
2023-05-15T21:10:52.3857345Z   Score:             19808 (max: 43043)
````

fbrausse commented:
Did you consider a lookup-based method, for `atol` e.g.
```c
/* one plus the numeric value, rest is zero to also catch '\0' */
static const unsigned char MAP[256] = {
	['0'] = 1, ['1'] = 2, ['2'] = 3, ['3'] = 4, ['4'] = 5,
	['5'] = 6, ['6'] = 7, ['7'] = 8, ['8'] = 9, ['9'] = 10,
};
```

As far as I know also the ctype functions like `isdigit`, etc. are usually implemented with lookup tables.

lucasccordeiro commented:
Thanks, @fbrausse. This is a good suggestion.

@kunjsong01 has volunteered to take over this PR.

@kunjsong01: @fbrausse suggestion should be the next step ;-)

lucasccordeiro commented:
If we consider 300 seconds.

Master:
````
Statistics:          26459 Files
   correct:           15604
     correct true:     8882
     correct false:    6722
   incorrect:           97
     incorrect true:     43
     incorrect false:    54
   unknown:           10758
   Score:             22246 (max: 43043)
````

This PR:
````
2023-05-19T11:53:50.9490668Z Statistics:          26459 Files
2023-05-19T11:53:50.9513061Z   correct:           15600
2023-05-19T11:53:50.9513893Z     correct true:     8877
2023-05-19T11:53:50.9514527Z     correct false:    6723
2023-05-19T11:53:50.9515545Z   incorrect:            97
2023-05-19T11:53:50.9516079Z     incorrect true:     43
2023-05-19T11:53:50.9516581Z     incorrect false:    54
2023-05-19T11:53:50.9517096Z   unknown:           10762
2023-05-19T11:53:50.9517818Z   Score:             22237 (max: 43043)
````
